File: 347. Top K Frequent Elements: O(nlogn).py
# 347. Top K Frequent Elements
# https://leetcode.com/problems/top-k-frequent-elements/description/

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        map = {}

        # Creates a hashmap how many instances each number has:
        for i in range(len(nums)):
            map[nums[i]] = map.get(nums[i], 0) + 1

        # Sorts key-value pairs by values (biggest to lowest) and returns array of k largest values:
        sorted_items = sorted(map.items(), key=lambda x: x[1], reverse=True)[:k]
        
        # Returns array of keys from these top k key-value pairs:
        keys = [item[0] for item in sorted_items]
        
        return keys

File: 347. Top K Frequent Elements: O(nlogn).rb
# 347. Top K Frequent Elements
# https://leetcode.com/problems/top-k-frequent-elements/description/

# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer[]}

# O(nlogn) algorithm for k-predominant numbers in array using built-in sorting on hashmap:
def top_k_frequent(nums, k)
    map = {}

    # Creates a hashmap how many instances each number has:
    nums.each do |num|
        map[num] ||= 0
        map[num] += 1
    end

    # Sorts key-value pairs by values (largest to smallest) and returns array of k largest values:
    sorted_items = map.sort_by { |key, value| -value }[0...k]

    # Returns array of keys from these top k key-value pairs:
    keys = sorted_items.map { |item| item[0] }
    keys
end
